### PR TITLE
MRG, FIX: Check reject_by_annotation in ica.plot_properties

### LINKED ISSUE #8101: ica.plot_properties error bug
#### Describe the bug
`ica.plot_properties` doesn't work for raw annotated with bad segments

#### Steps to reproduce
```python
import os
import mne
from mne.preprocessing import (ICA, create_eog_epochs, create_ecg_epochs,
                               corrmap)

sample_data_folder = mne.datasets.sample.data_path()
sample_data_raw_file = os.path.join(sample_data_folder, 'MEG', 'sample',
                                    'sample_audvis_raw.fif')
raw = mne.io.read_raw_fif(sample_data_raw_file)
raw.crop(tmax=60.)

a = mne.Annotations(onset=[3, 5, 7],
                           duration=[1, 0.5, 0.25],
                           description=['BAD', 'BAD', 'BAD'])
raw.set_annotations(a)

ica = mne.preprocessing.ICA()#max_pca_components=29)
ica.fit(raw)

ica.plot_properties(raw, picks=[0, 1])
```

#### Expected results
picks with component properties

#### Actual results
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-7-dae74f67a92b> in <module>
----> 1 ica.plot_properties(raw, picks=[0, 1])

/usr/local/anaconda3/lib/python3.7/site-packages/mne/preprocessing/ica.py in plot_properties(self, inst, picks, axes, dB, plot_std, topomap_args, image_args, psd_args, figsize, show, reject)
   1744                                    topomap_args=topomap_args,
   1745                                    image_args=image_args, psd_args=psd_args,
-> 1746                                    figsize=figsize, show=show, reject=reject)
   1747 
   1748     @copy_function_doc_to_method_doc(plot_ica_sources)

/usr/local/anaconda3/lib/python3.7/site-packages/mne/viz/ica.py in plot_ica_properties(ica, inst, picks, axes, dB, plot_std, topomap_args, image_args, psd_args, figsize, show, reject)
    404 
    405     epochs_src = ica.get_sources(inst_rejected)
--> 406     data = epochs_src.get_data()
    407 
    408     ica_data = np.swapaxes(data[:, picks, :], 0, 1)

/usr/local/anaconda3/lib/python3.7/site-packages/mne/epochs.py in get_data(self, picks, item)
   1440             A view on epochs data.
   1441         """
-> 1442         return self._get_data(picks=picks, item=item)
   1443 
   1444     @property

<decorator-gen-188> in _get_data(self, out, picks, item, verbose)

/usr/local/anaconda3/lib/python3.7/site-packages/mne/epochs.py in _get_data(self, out, picks, item, verbose)
   1367                     else:
   1368                         epoch_noproj = None
-> 1369                         epoch = self._data[idx]
   1370                 else:  # from disk
   1371                     epoch_noproj = self._get_epoch_from_raw(idx)

IndexError: index 27 is out of bounds for axis 0 with size 27〉
```

#### Additional information
Platform:      Linux-4.18.0-25-generic-x86_64-with-debian-buster-sid
Python:        3.7.7 (default, Mar 26 2020, 15:48:22)  [GCC 7.3.0]
Executable:    /usr/local/anaconda3/bin/python
CPU:           x86_64: 8 cores
Memory:        31.3 GB

mne:           0.21.dev0
numpy:         1.18.1 {blas=mkl_rt, lapack=mkl_rt}
scipy:         1.4.1
matplotlib:    3.1.3 {backend=module://ipykernel.pylab.backend_inline}

sklearn:       0.22.1
numba:         0.48.0
nibabel:       Not found
cupy:          Not found
pandas:        1.0.3
dipy:          Not found
mayavi:        Not found
pyvista:       Not found
vtk:           Not found

### PR DESCRIPTION
**Pull request by yh-luo**
Fix #8101 .

Because `reject_by_annotation` in `ica.fit()` defaults to `True`, reconstruction of source in `plot_properties` should take care of that or the error occurs due to the incorrect number of epochs.

**TODO**
- _Fix `ica.plot_properties`_
- _Fix `make_fixed_length_epochs`_
- _Update documentation of `reject_by_annotation` using docdict_
- _Update what's new_

### Before this PR

After adding annotations
```python
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-d42215032b75> in <module>
      6 
      7 ica.fit(raw)
----> 8 ica.plot_properties(raw, picks=[0, 1])

~/env/mne/lib64/python3.7/site-packages/mne/preprocessing/ica.py in plot_properties(self, inst, picks, axes, dB, plot_std, topomap_args, image_args, psd_args, figsize, show, reject)
   1631                                    topomap_args=topomap_args,
   1632                                    image_args=image_args, psd_args=psd_args,
-> 1633                                    figsize=figsize, show=show, reject=reject)
   1634 
   1635     @copy_function_doc_to_method_doc(plot_ica_sources)

~/env/mne/lib64/python3.7/site-packages/mne/viz/ica.py in plot_ica_properties(ica, inst, picks, axes, dB, plot_std, topomap_args, image_args, psd_args, figsize, show, reject)
    403 
    404     epochs_src = ica.get_sources(inst_rejected)
--> 405     data = epochs_src.get_data()
    406 
    407     ica_data = np.swapaxes(data[:, picks, :], 0, 1)

~/env/mne/lib64/python3.7/site-packages/mne/epochs.py in get_data(self, picks, item)
   1412             A view on epochs data.
   1413         """
-> 1414         return self._get_data(picks=picks, item=item)
   1415 
   1416     @property

<decorator-gen-182> in _get_data(self, out, picks, item, verbose)

~/env/mne/lib64/python3.7/site-packages/mne/epochs.py in _get_data(self, out, picks, item, verbose)
   1343                     else:
   1344                         epoch_noproj = None
-> 1345                         epoch = self._data[idx]
   1346                 else:  # from disk
   1347                     epoch_noproj = self._get_epoch_from_raw(idx)

IndexError: index 28 is out of bounds for axis 0 with size 28
```
Figure_1

---

### After this PR

Before adding annotations
```python
import os

import mne

sample_data_folder = mne.datasets.sample.data_path()
sample_data_raw_file = os.path.join(sample_data_folder, 'MEG', 'sample',
                                    'sample_audvis_raw.fif')
raw = mne.io.read_raw_fif(sample_data_raw_file)
raw.crop(tmax=60)

raw.load_data()
raw.filter(1, 40, n_jobs='cuda')

ica = mne.preprocessing.ICA()

ica.fit(raw)
ica.plot_properties(raw, picks=[0, 1])
```
image
image

After adding annotations
```python

# refitting with annotations
a = mne.Annotations(onset=[3],
                    duration=[1],
                    description=['BAD'])
raw.set_annotations(a)

ica.fit(raw)
ica.plot_properties(raw, picks=[0, 1])
```
image
image

This PR fixes the bug and the results are not influenced.

### CODE REVIEW

**Comment by agramfort** [comment]
can you update  or add a test @yh-luo ?

**Comment by yh-luo** [comment]
@agramfort I will work on it!

**Comment by yh-luo** [comment]
After some digging, the bug is actually caused by `mne.epochs.make_fixed_length_epochs`. This PR added test for it too.

File: mne/epochs.py [hunk 1]
```diff
@@ -3288,6 +3288,12 @@ def make_fixed_length_epochs(raw, duration=1.,
     duration : float
         Duration of each epoch in seconds. Defaults to 1.
     %(preload)s
+    reject_by_annotation : bool
+        Whether to reject based on annotations. If ``True`` (default), epochs
+        overlapping with segments whose description begins with ``'bad'`` are
+        rejected. If ``False``, no rejection based on annotations is performed.
```
Comment (agramfort) [inline mne/epochs.py:3294]: can you factorize this docstring using docdict to avoid a copy paste?

File: mne/viz/ica.py [hunk 2]
```diff
@@ -309,6 +310,12 @@ def plot_ica_properties(ica, inst, picks=None, axes=None, dB=True,
         If None, no rejection is applied. The default is 'auto',
         which applies the rejection parameters used when fitting
         the ICA object.
+    reject_by_annotation : bool
+        Whether to omit bad segments from the data. If ``True`` (default),
+        annotated segments with a description that starts with ``'bad'`` are
+        omitted. Has no effect if ``inst`` is an Epochs object.
```
Comment (agramfort) [inline mne/viz/ica.py:316]: here too

**Comment by larsoner** [comment]
CircleCI can be ignored, it's fixed by #8109 (will restart once that's merged)

File: mne/epochs.py (hunk 1 above)
Comment (yh-luo) [reply mne/epochs.py]: A lot of functions duplicate the documentation of `reject_by_annotation`. I updated most of them too (except those default to `False`).

**Comment by agramfort** [comment]
@yh-luo please add a what's new entry and then +1 for MRG. thx

**Comment by larsoner** [comment]
Thanks @yh-luo !

**Pull request merged**